include/up-cpp/datamodel/builder/UMessage.h

```diff
@@ -0,0 +1,276 @@
+	/// will be used.
+	///
+	/// @post The default message priority will be replaced with the
+	///       provided priority for all future generated messages.
+	///
+	/// @param The priority to use when building messages.
+	///
+	/// @throws std::out_of_range when setting a priority lower than CS4
+	///                           for "request" or "response" messages.
+	/// @returns A reference to this UMessageBuilder
+	UMessageBuilder& withPriority(v1::UPriority);
+
+	/// @brief Set the ttl to use for building messages.
+	///
+	/// If not set, then no TTL will be set in the built message.
+	///
+	/// @post All future generated messages will contain the new TTL value.
+	///
+	/// @param Amount of time from the call to build() that a message will
+	///        still be considered valid. Must be greater than 0 and fit
+	///        within a uint32_t.
+	///
+	/// @throws std::out_of_range if (ttl <= 0) or (ttl > uint32_t max).
+	/// @returns A reference to this UMessageBuilder
+	UMessageBuilder& withTtl(std::chrono::milliseconds);
+
+	/// @brief Set the message's authorization token used for TAP.
+	///
+	/// If not set, then no token will be set in the built message.
+	///
+	/// @note Authorization tokens are only used for RPC request messages.
+	/// @post All future generated messages will contain the new token.
+	///
+	/// @param Authorization token to use for built messages.
+	///
+	/// @throws std::domain_error when called on a message with any type
+	///                           other than "request".
+	/// @returns A reference to this UMessageBuilder
+	UMessageBuilder& withToken(const std::string&);
+
+	/// @brief Set the message's authorization level.
+	///
+	/// If not set, then no auth level will be set in the built message.
+	///
+	/// @note Authorization levels are only used for RPC request messages.
+	/// @post All future generated messages will contain the new
+	///       authorization level.
+	///
+	/// @param The permission level to use for built messages.
+	///
+	/// @throws std::domain_error when called on a message with any type
+	///                           other than "request".
+	/// @returns A reference to this UMessageBuilder
+	UMessageBuilder& withPermissionLevel(int32_t);
+
+	/// @brief Sets the response's commstatus field indicating an error
+	///        occurred with the request or in generating the response.
+	///
+	/// If not set, then no commstatus will be set in the built message.
+	///
+	/// @note Communication statuses are only used for RPC response
+	///       messages.
+	/// @post All future generated messages will contain the new
+	///       commstatus code.
+	///
+	/// @param A UCode indicating the error. If the value is OK, the
+	///        message will send no commstatus.
+	///
+	/// @throws std::domain_error when called on a message with any type
+	///                           other than "response".
+	/// @returns A reference to this UMessageBuilder
+	UMessageBuilder& withCommStatus(v1::UCode);
```

Comment:
Sink and request ID are populated in the appropriate static methods and can't be overridden after the fact. When they are needed, they are not optional.

Reply:
What exactly is the usecase for `withTransparent()`? It doesn't seem that it would be used in any of the communication modules, and the documentation only says that it's for cloudevent tracing. In what scenarios would it be used and set? I think I am inclined to hold off on adding it until we have a known use for it.